crates/red_knot_python_semantic/resources/mdtest/boolean/short_circuit.md

```diff
@@ -0,0 +1,77 @@
+# Short-Circuit Evaluation
+
+## Not all boolean expressions must be evaluated
+
+In `or` expressions, if the left-hand side is truthy, the right-hand side is not evaluated.
+Similarly, in `and` expressions, if the left-hand side is falsy, the right-hand side is not evaluated.
+
+```py
+def bool_instance() -> bool:
+    return True
+
+if bool_instance() or (x := 1):
+    # error: "Name `x` used when possibly not defined"
```

Comment:
nit: it's better to assert on the error code here rather than the error message, so that if we change the message in the future the test still passes (same comment for a few other places in this PR). It _would_ be good to assert on the error message if this was a place where it was ambiguous which variable the error was being emitted about, but here it's not :-)

```suggestion
    # error: [possibly-unresolved-reference]
```